src/audio/host.c

```diff
@@ -226,6 +230,140 @@ static void host_dma_cb(void *arg, enum notify_id type, void *data)
 		host_one_shot_cb(dev, bytes);
 }
 
+static uint32_t host_get_copy_bytes_one_shot(struct comp_dev *dev)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+	struct dma_sg_elem *local_elem = hd->config.elem_array.elems;
+	uint32_t copy_bytes = 0;
+	uint32_t split_value;
+	uint32_t flags = 0;
+
+	buffer_lock(hd->local_buffer, &flags);
+
+	/* calculate minimum size to copy */
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+		copy_bytes = hd->local_buffer->stream.free;
+	else
+		copy_bytes = hd->local_buffer->stream.avail;
+
+	buffer_unlock(hd->local_buffer, flags);
+
+	/* copy_bytes should be aligned to minimum possible chunk of
+	 * data to be copied by dma.
+	 */
+	copy_bytes = ALIGN_DOWN(copy_bytes, hd->dma_copy_align);
+
+	split_value = host_dma_get_split(hd, copy_bytes);
+	if (split_value)
+		copy_bytes -= split_value;
+
+	local_elem->size = copy_bytes;
+
+	return copy_bytes;
+}
+
+static int host_copy_one_shot(struct comp_dev *dev)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+	uint32_t copy_bytes = 0;
+	int ret = 0;
+
+	comp_dbg(dev, "host_copy_one_shot()");
+
+	/* update first transfer manually */
+	if (!dev->position)
+		host_one_shot_cb(dev, hd->dma_buffer->stream.size);
+
+	copy_bytes = host_get_copy_bytes_one_shot(dev);
+	if (!copy_bytes) {
+		comp_info(dev, "host_copy_one_shot(): no bytes to copy");
+		return ret;
+	}
+
+	/* reconfigure transfer */
+	ret = dma_set_config(hd->chan, &hd->config);
+	if (ret < 0) {
+		comp_cl_err(&comp_host, "host_copy_one_shot(): dma_set_config() failed, ret = %u",
+			    ret);
+		return ret;
+	}
+
+	ret = dma_copy(hd->chan, copy_bytes, DMA_COPY_ONE_SHOT);
+	if (ret < 0) {
+		comp_cl_err(&comp_host, "host_copy_one_shot(): dma_copy() failed, ret = %u",
+			    ret);
+		return ret;
+	}
+
+	return ret;
+}
+
+static uint32_t host_get_copy_bytes_normal(struct comp_dev *dev)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+	uint32_t avail_bytes = 0;
+	uint32_t free_bytes = 0;
+	uint32_t copy_bytes = 0;
+	uint32_t flags = 0;
+	int ret;
+
+	/* get data sizes from DMA */
+	ret = dma_get_data_size(hd->chan, &avail_bytes,
+				&free_bytes);
+	if (ret < 0) {
+		comp_cl_err(&comp_host, "host_get_copy_bytes_normal(): dma_get_data_size() failed, ret = %u",
+			    ret);
+		return 0;
+	}
+
+	buffer_lock(hd->local_buffer, &flags);
+
+	/* calculate minimum size to copy */
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+		copy_bytes = MIN(avail_bytes,
+				 hd->local_buffer->stream.free);
+	else
+		copy_bytes = MIN(hd->local_buffer->stream.avail,
+				 free_bytes);
+
+	buffer_unlock(hd->local_buffer, flags);
+
+	/* copy_bytes should be aligned to minimum possible chunk of
+	 * data to be copied by dma.
+	 */
+	copy_bytes = ALIGN_DOWN(copy_bytes, hd->dma_copy_align);
+
+	return copy_bytes;
+}
+
+static int host_copy_normal(struct comp_dev *dev)
```

Comment:
Could we have a one line comment above each function describing the copy use case

Reply:
Done.